src/main/java/de/tum/cit/aet/artemis/programming/service/localci/scaparser/strategy/sarif/SarifParser.java

```diff
@@ -0,0 +1,142 @@
+        PhysicalLocation location = result.getLocations().flatMap(locations -> locations.stream().findFirst()).flatMap(Location::getPhysicalLocation)
+                .orElseThrow(() -> new RuntimeException("Location needed"));
+
+        URI uri = URI.create(location.getArtifactLocation().flatMap(ArtifactLocation::getUri).orElseThrow(() -> new RuntimeException("File path needed")));
+        String path = uri.getPath();
+
+        Region region = location.getRegion().orElseThrow(() -> new RuntimeException("Region must be present"));
+        int startLine = region.getStartLine().orElseThrow(() -> new RuntimeException("Text region needed"));
+        int startColumn = region.getStartColumn().orElse(1);
+        int endLine = region.getEndLine().orElse(startLine);
+        int endColumn = region.getEndColumn().orElse(startColumn + 1);
+
+        String ruleId = getRuleId(result);
+
+        Optional<Integer> ruleIndex = getRuleIndex(result);
+
+        Optional<ReportingDescriptor> ruleByIndex = driver.getRules().flatMap(rules -> ruleIndex.map(rules::get));
+        Optional<ReportingDescriptor> rule = ruleByIndex.or(() -> lookupRuleById(ruleId, ruleOfId));
+
+        // Fallback to the rule identifier for the category
+        String category = rule.map(ruleCategorizer::categorizeRule).orElse(ruleId);
+
+        Result.Level level = result.getLevel().orElse(Result.Level.WARNING);
+
+        String message = findMessage(result, driver, rule);
+
+        return new StaticCodeAnalysisIssue(path, startLine, endLine, startColumn, endColumn, ruleId, category, message, level.toString(), null);
+    }
+
+    private static String getRuleId(Result result) {
+        return result.getRuleId()
+                .orElseGet(() -> result.getRule().flatMap(ReportingDescriptorReference::getId).orElseThrow(() -> new RuntimeException("Either ruleId or rule.id must be present")));
+    }
+
+    private static Optional<Integer> getRuleIndex(Result result) {
+        // ruleIndex can use -1 to indicate a missing value
+        Optional<Integer> ruleIndexOrMinusOne = result.getRuleIndex().or(() -> result.getRule().flatMap(ReportingDescriptorReference::getIndex));
+        return ruleIndexOrMinusOne.flatMap(index -> index != -1 ? Optional.of(index) : Optional.empty());
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private static String findMessage(Result result, ToolComponent driver, Optional<ReportingDescriptor> rule) {
+        return result.getMessage().getText().orElseGet(() -> lookupMessageById(result, driver, rule));
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private static String lookupMessageById(Result result, ToolComponent driver, Optional<ReportingDescriptor> rule) {
+        String messageId = result.getMessage().getId().orElseThrow(() -> new RuntimeException("Either text or id must be present"));
```

Comment:
Should warnings where no message can be found just be skipped instead? The result as a whole might still have other relevant warnings.